units/generate_tasks.go

```diff
@@ -50,75 +46,99 @@ func makeGenerateTaskJob() *generateTasksJob {
 	return j
 }
 
-func NewGenerateTasksJob(id string, json []json.RawMessage) amboy.Job {
+func NewGenerateTasksJob(id string, ts string) amboy.Job {
 	j := makeGenerateTaskJob()
 	j.TaskID = id
-	j.JSON = json
 
-	j.SetID(fmt.Sprintf("%s-%s", generateTasksJobName, id))
+	j.SetID(fmt.Sprintf("%s-%s-%s", generateTasksJobName, id, ts))
 	return j
 }
 
+func (j *generateTasksJob) generate(ctx context.Context, t *task.Task) error {
+	t, err := task.FindOneId(j.TaskID)
+	if err != nil {
+		return errors.Wrapf(err, "problem finding task %s", j.TaskID)
+	}
+	if t == nil {
+		return errors.Errorf("task %s does not exist", j.TaskID)
+	}
+	if t.GeneratedTasks {
+		grip.Debug(message.Fields{
+			"message": "attempted to generate tasks, but generator already ran for this task",
+			"task":    t.Id,
+			"version": t.Version,
+		})
+		return nil
+	}
+
+	projects, err := parseProjects(t.GeneratedJSON)
+	if err != nil {
+		return errors.Wrap(err, "error parsing JSON from `generate.tasks`")
+	}
+	g := model.MergeGeneratedProjects(projects)
+	g.TaskID = j.TaskID
+
+	p, v, t, pm, err := g.NewVersion()
+	if err != nil {
+		return errors.Wrap(err, "problem creating new version")
+	}
+	if err = validator.CheckProjectConfigurationIsValid(p); err != nil {
+		return errors.Wrap(err, "project configuration was invalid")
+	}
+
+	// Don't use the job's context, because it's better to finish than to exit early after a
+	// SIGTERM from a deploy. This should maybe be a context with timeout.
+	err = g.Save(context.TODO(), p, v, t, pm)
```

Comment:
make this a background? 

I don't think respecting the context is enough because we need something in the queue infra that will block while waiting for the queue to close, which may or may not exist.